src/core/algorithm/graphToMarkovChain.test.js

```diff
@@ -122,6 +122,31 @@ describe("core/algorithm/graphToMarkovChain", () => {
         MapUtil.mapValues(map, (_, v) => sortBy(v, (x) => JSON.stringify(x)));
       expect(canonicalize(actual)).toEqual(canonicalize(expected));
     });
+    it("works on a chain with loops", () => {
+      const loop = edge("loop", n1, n1);
+      const e1 = edge("e1", n1, n2);
+
+      const g = new Graph().addNode(n1).addNode(n2).addEdge(loop).addEdge(e1);
+      const edgeWeight = () => ({forwards: 6.0, backwards: 3.0});
+      const actual = createConnections(g, edgeWeight, 1.0);
+      // Total out-weights (for normalization factors):
+      //   - for `n1`: 2 out, 1 in, 1 synthetic: 12 + 3 + 1 = 16
+      //   - for `n2`: 0 out, 1 in, 1 synthetic: 0 + 3 + 1 = 7
+      const expected = new Map()
+        .set(n1.address, [
+          {adjacency: {type: "SYNTHETIC_LOOP"}, weight: 1 / 16},
+          {adjacency: {type: "IN_EDGE", edge: loop}, weight: 6 / 16},
+          {adjacency: {type: "OUT_EDGE", edge: loop}, weight: 3 / 16},
+          {adjacency: {type: "OUT_EDGE", edge: e1}, weight: 3 / 4},
+        ])
+        .set(n2.address, [
+          {adjacency: {type: "SYNTHETIC_LOOP"}, weight: 1 / 4},
+          {adjacency: {type: "IN_EDGE", edge: e1}, weight: 6 / 16},
+        ]);
```

Comment:
What was not obvious here, why the normalized weight groups are spread out across different map entries.
Highlighting as green and red pairs. 

```diff
      const expected = new Map()
        .set(n1.address, [
+         {adjacency: {type: "SYNTHETIC_LOOP"}, weight: 1 / 16},
+         {adjacency: {type: "IN_EDGE", edge: loop}, weight: 6 / 16},
+         {adjacency: {type: "OUT_EDGE", edge: loop}, weight: 3 / 16},
-         {adjacency: {type: "OUT_EDGE", edge: e1}, weight: 3 / 4},
        ])
        .set(n2.address, [
-         {adjacency: {type: "SYNTHETIC_LOOP"}, weight: 1 / 4},
+         {adjacency: {type: "IN_EDGE", edge: e1}, weight: 6 / 16},
        ]);
```

Reading through the implementation code (not the point of documentation :slightly_smiling_face:), finding this representation stores the weight as normalized on the other end of this edge.

So for a `n1 <-> n2` edge, in `n1`'s entry will be normalized to the `n2` total weights.

No idea why, but that appears to be intentional.

Reply:
It’s certainly intentional. [The documentation][1] says, 

> This `weight` is a conditional probability: given that you're at the
> source of this connection's adjacency, what's the probability that you
> travel along this connection to the target?

That is, these are `Pr[X_{n+1} = dst | X_n = src]`, so the weights must
be normalized per source to form a proper probability distribution. When
you say “a `n1 <-> n2` edge”, remember that this really represents two
edges, `e1: n1 -> n2` and `e1': n2 -> n1`. So there are two adjacencies:
the adjacency for `e1` is stored in the map entry for `n2`, and is
normalized by the total out-weight of its source, `n1`; and the
adjacency for `e1'` is stored in the map entry for `n1`, and is
normalized by the total out-weight of its source, `n2`.

[1]: https://github.com/sourcecred/sourcecred/blob/ae181c2fda8719546d90acbaa0ab60d417473a2c/src/core/algorithm/graphToMarkovChain.js#L16-L18